src/Control/Carrier/State/IORef.hs

```diff
@@ -0,0 +1,97 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{- | A carrier for the 'State' effect. It uses an 'IORef' internally to handle its state, and thus admits a 'MonadUnliftIO' instance. Because the state operations are performed impurely, this carrier will not lose state effects even with nefarious uses of 'liftWith'.
+
+Unlike the other carriers for 'State', this carrier's effects will not backtrack when run in conjuction with 'NonDet' effects.
+
+@since 1.1.2.0
+-}
+module Control.Carrier.State.IORef
+( -- * Impure state carrier
+  runState
+, runStateRef
+, evalState
+, execState
+, StateC(..)
+-- * State effect
+, module Control.Effect.State
+) where
+
+import           Control.Algebra
+import           Control.Applicative (Alternative(..))
+import           Control.Carrier.Reader
+import           Control.Effect.State
+import           Control.Monad (MonadPlus(..))
+import qualified Control.Monad.Fail as Fail
+import           Control.Monad.Fix
+import           Control.Monad.IO.Class
+import           Control.Monad.IO.Unlift
+import           Control.Monad.Trans.Class
+import           Data.IORef
+
+-- | Run a 'State' effect starting from the passed value.
+--
+--   prop> run (runState a (pure b)) === (a, b)
+--
+-- @since 1.1.2.0
+runState :: MonadIO m => s -> StateC s m a -> m (s, a)
+runState s x = do
+  ref <- liftIO $ newIORef s
+  result <- runReader ref . runStateC $ x
+  final <- liftIO . readIORef $ ref
+  pure (final, result)
+{-# INLINE[3] runState #-}
+
+-- | Run a 'State' effect starting from the passed 'IORef'.
+--
+--   prop> run (runState a (pure b)) === (a, b)
```

Comment:
1. See earlier note re: API doc format.
2. This is the wrong property for this function.